caliban_toolbox/pre_annotation/data_loader.py

```diff
@@ -0,0 +1,462 @@
+# & National Institutes of Health (NIH) under Grant U24CA224309-01.
+# All rights reserved.
+#
+# Licensed under a modified Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.github.com/vanvalenlab/caliban-toolbox/LICENSE
+#
+# The Work provided may be used for non-commercial academic purposes only.
+# For any other use of the Work, including commercial use, please contact:
+# vanvalenlab@gmail.com
+#
+# Neither the name of Caltech nor the names of its contributors may be used
+# to endorse or promote products derived from this software without specific
+# prior written permission.
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Load raw data (with metadata) from CellNet-formatted ontology."""
+
+from __future__ import absolute_import
+
+import os
+import json
+import fnmatch
+
+from pathlib import Path
+import numpy as np
+
+from skimage.external import tifffile as tiff
+# from pymongo import MongoClient
+import pandas as pd
+
+from caliban_toolbox.utils.misc_utils import sorted_nicely
+
+
+class UniversalDataLoader(object):
+    """Given a CellNet data type, load and store image set and metadata for
+       initial predictions and crowdsourcing annotation curation.
+
+    The raw data and metadata file should be arranged according to the CellNet
+    data ontology. The root of this ontology should be mounted as /data within
+    the container. The image files are stored in a numpy array together with a
+    dictionary object for the metadata.
+
+    Excluding data and imaging type, Arg options include: all and random
+    (random picks one file at random - best used for testing).
+
+    Args:
+        data type (list): CellNet data type ('dynamic/static', '2d/3d')
+        imaging types (list): imaging modality of interest ('fluo', 'phase', etc)
+        specimen types (list): specimen of interest (HEK293, HeLa, etc)
+        compartments (list): compartments of interest (nuclear, whole_cell)
+        marker (list): marker of interest
+        exp_ids/DOIs (list): Experiment ID or DOI of the dataset
+        sessions (list): which sessionss to include
+        positions/FOVs (list): which positionss/FOVs to include
+
+            - list should be strings and match CellNet ontology
+              (e.g. data type = ['dynamic', '2d'])
+            - 'all' selects all data from a given catagory
+              (e.g. sessions=['all'])
+
+    Returns:
+        Numpy array with the shape [fovs, tifs, y_dim, x_dim]
```

Comment:
Is tifs the same as markers/channels?

Reply:
I don't know what you mean by markers/channels, you'll need to be more specific. Our ontology stores images according to static/dynamic, 2d/3d, imaging type, specimen type, compartment_marker (eg. nuclear_dapi), and experiment ID. The files in this final directory are expected to be named according to session and FOV/position. So there are no "marker/channels" to return. The metadata files keeps a record of whats in each row of the array for reference.

Reply:
Ah, sorry. My question is what doe the TIFs axis represent here. Is this whatever extra dimension the data has? For example, for 3D data, is TIFs the z-dimension, and there isn't a channels dimension, whereas for multichannel data TIFs is the channels dimension?

Reply:
There is no multichannel data for the function to load, as that is handled by virtue of the ontology - each channel has its own sub directory.  

I have amended the docstring to make it clear that the tifs dim is a z-dim that either represents space or time. This is temporary as we will need to accommodate 3D dynamic data as well, but that is easily solved with another sql lookup and dimension.